zaza/model.py

```diff
@@ -1320,12 +1330,31 @@ async def async_wait_for_application_states(model_name=None, states=None,
             timed_out = int(time.time() - start) > timeout
             issues = []
             for application in applications_left.copy():
+                check_info = states.get(application, {})
                 app_data = model.applications.get(application, None)
                 units = list(app_data.units)
-                # if there are no units the application is 'ready'
-                if len(units) > 0:
-                    check_info = states.get(application, {})
 
+                # all_okay is a Boolean of the current state.  It starts as
+                # True, but if False by the end of the checks, then the
+                # application is not ready.
+                all_okay = True
+
+                # if there are no units then the application may not be ready.
+                # However, if the caller explicitly allows that situation then
+                # we gate on that.
+                num_expected = check_info.get('num-expected-units', None)
+                if num_expected is not None:
+                    if len(units) != num_expected:
+                        print("all_okay is now false")
```

Comment:
Debug

Reply:
Yes, I'll remove; funny how these tend to creep in :)